README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->
<img align="right" src="figs/icon-1.png" style="max-width: 300px">

mandelbrot
==========

[![travis\_status](https://travis-ci.org/blmoore/mandelbrot.svg?branch=master)](https://travis-ci.org/blmoore/mandelbrot) [![codecov](https://codecov.io/gh/blmoore/mandelbrot/branch/master/graph/badge.svg)](https://codecov.io/gh/blmoore/mandelbrot) [![CRAN\_badge](http://www.r-pkg.org/badges/version/mandelbrot)](https://cran.r-project.org/package=mandelbrot)

Curious whether your favourite complex number is a member of the Mandelbrot set?

—No? Then probably you just want to draw pretty pictures.

Install
-------

Install from CRAN:

``` r
install.packages("mandelbrot")
```

Or get the development version:

``` r
devtools::install_github("blmoore/mandelbrot")
```

Examples
--------

``` r
library(mandelbrot)

mb <- mandelbrot()
plot(mb)
```

![](figs/README-b_n_w-1.png)

Pretty colours:

``` r
mb <- mandelbrot(xlim = c(-0.8438146, -0.8226294),
                 ylim = c(0.1963144, 0.2174996), 
                 iterations = 500, resolution = 800)

cols <- mandelbrot_palette(RColorBrewer::brewer.pal(11, "Spectral"), fold = FALSE)
plot(mb, col = cols)
```

![](figs/README-trip-1.png)

Experiment with transforms:

``` r
# layout config (reset after)
par(mfrow = c(1, 2), pty = "s", mar = rep(0, 4))
plot(mb, col = cols, transform = "inverse")
plot(mb, col = cols, transform = "log")
```

![](figs/README-trans-1.png)

Don't like `image`? Convert to a `data.frame` and use with ggplot2:

``` r
library(ggplot2)

mb <- mandelbrot(xlim = c(-0.8335, -0.8325),
                 ylim = c(0.205, 0.206), 
                 resolution = 1200L,
                 iterations = 1000)

# weird uneven palette made for a heatmap viz:
# https://benjaminlmoore.wordpress.com/2015/04/09/recreating-the-vaccination-heatmaps-in-r/
cols <- c(
  colorRampPalette(c("#e7f0fa", "#c9e2f6", "#95cbee",
                     "#0099dc", "#4ab04a", "#ffd73e"))(10),
  colorRampPalette(c("#eec73a", "#e29421", "#e29421", 
                     "#f05336","#ce472e"), bias=2)(90), 
  "black")

df <- as.data.frame(mb)
ggplot(df, aes(x = x, y = y, fill = value)) +
  geom_raster(interpolate = TRUE) + theme_void() +
  scale_fill_gradientn(colours = cols, guide = "none") 
```

![](figs/README-ggplot-1.png)

`mandelbrot0` is a faster interface for `as.data.frame(mandelbrot(...))`.

See also
--------

[shinybrot](https://github.com/blmoore/shinybrot) is a Shiny web app that enables interactive exploration of the Mandelbrot set as generated by this package.

Hosted on shinyapps: <https://blmr.shinyapps.io/shinybrot>

Credits
-------

-   Wraps original R / C code by [mariodosreis](https://github.com/mariodosreis) (see R package [fractal](https://github.com/mariodosreis/fractal)).


NEWS.md
# mandelbrot 0.2.0

* Added a `NEWS.md` file to track changes to the package.
* Uploaded to CRAN w/ CRAN related fixes
* Tweaked interfaces for `plot.mandelbrot` and `mandelbrot_palette`

# mandelbrot 0.1.0

* First release





tests/testthat/test_main.R
context("Mandelbrot interfaces")

test_that("mandebrot generates correct results", {

  # range in set
  for (iter in c(10, 100, 250)) {
    mb <- as.data.frame(mandelbrot(iterations = iter))

    # test point <0, 0>
    x0y0 <- mb[mb$y > -.01 & mb$y < .01 & mb$x > -.01 & mb$x < .01,]

    # iter +1 == never goes to inf / >2
    expect_equal(unique(x0y0$value), iter + 1)
  }

  # range outside set
  for (iter in c(10, 100, 250)) {
    mb <- as.data.frame(mandelbrot(iterations = iter))

    # test point <1, 1>
    x1y1 <- mb[mb$y > .99 & mb$y < 1.01 & mb$x > .99 & mb$x < 1.01,]

    # not in the set + goes >2 at 2
    expect_lt(unique(x1y1$value), iter)
    expect_equal(unique(x1y1$value), 2)
  }

})



R/wrapper.R

#' Calculate the Mandelbrot set
#'
#' Generates a view on the Mandelbrot set
#' using an underlying C function.
#'
#' \code{mandelbrot0} is an experimental interface
#' for generating tidy data.frames faster than
#' \code{as.data.frame(mandelbrot())}.
#'
#' @section Mandelbrot set:
#' In brief, the Mandelbrot set contains the complex numbers
#' where the 0 orbit of the following function remains
#' bounded (\eqn{<2}):
#' \deqn{f_{z+1} = z^2 + c}
#'
#' For information and discussion on the Mandelbrot and
#' related sets, one great resource is
#' \href{https://plus.maths.org/content/unveiling-mandelbrot-set}{plus.maths.org}.
#' There's also a popular
#' \href{https://www.youtube.com/watch?v=NGMRB4O922I}{YouTube video by Numberphile}.
#'
#' @section Credits:
#' Wraps original C code by Mario dos Reis, September 2003.
#'
#' @param xlim limits of x axis (real part)
#' @param ylim limits of y axis (imaginary part)
#' @param resolution either an integer \eqn{n} for \eqn{n^2} pixels
#'   or a list with x and y components specifying the resolution
#'   in each direction (e.g. \code{list(x = 500, y = 500)})
#' @param iterations maximum number of iterations to
#'   evaluate each case
#'
#' @return a \code{mandelbrot} structure with components: \code{x} a vector
#'   of the real parts of the x-axis; \code{y} the imaginary parts of each
#'   number (the y-axis); \code{z} a matrix of the number of iterations that
#'   \eqn{|z|<2}
#'
#' @references
#'   \url{https://stat.ethz.ch/pipermail/r-help/2003-October/039773.html}
#'   \url{http://people.cryst.bbk.ac.uk/~fdosr01/Rfractals/index.html}
#'
#' @useDynLib mandelbrot mandelbrot_
#'
#' @export
mandelbrot <- function(xlim = c(-2, 2), ylim = c(-2, 2),
  resolution = 600, iterations = 50) {

  if (is.list(xlim)) {
    ylim <- range(xlim$y)
    xlim <- range(xlim$x)
  }

  if (is.list(resolution)) {
    if (!length(resolution) == 2 | !all(c("x", "y") %in% names(resolution))) {
      stop("resolution should be a named list, e.g. list(x = 500, y = 500)")
    } else {
      x_res <- resolution$x
      y_res <- resolution$y
    }
  } else {
    if (is.numeric(resolution) & length(resolution) == 1) {
      resolution <- as.integer(resolution)
      x_res <- resolution
      y_res <- resolution
    } else {
      stop("resolution should be an integer, not ", resolution)
    }
  }

  if (!is.numeric(xlim) | !is.numeric(ylim)) {
    stop("xlim and ylim must be numeric")
  }

  x_coord <- seq(xlim[1], xlim[2], len = x_res)
  y_coord <- seq(ylim[1], ylim[2], len = y_res)
  set <- numeric(x_res * y_res)

  # This is the call to the C function itself
  this_set <- .C("mandelbrot_",
    xcoo = as.double(x_coord),
    ycoo = as.double(y_coord),
    nx = as.integer(x_res),
    ny = as.integer(y_res),
    set = as.integer(set),
    iter = as.integer(iterations))$set

  # Create a list with elements x, y and z,
  # suitable for image(), persp(), etc. and return it.
  structure(
    list(x = x_coord, y = y_coord,
      z = matrix(this_set, ncol = y_res, byrow = T)
    ),
    class = c("mandelbrot", "list")
  )
}

#' @rdname mandelbrot
#'
#' @useDynLib mandelbrot mandelbrot_alt
#'
#' @export
mandelbrot0 <- function(xlim = c(-2, 2), ylim = c(-2, 2),
  resolution = 600, iterations = 50) {

  if (is.list(xlim)) {
    ylim <- range(xlim$y)
    xlim <- range(xlim$x)
  }

  if (is.list(resolution)) {
    stop("mandelbrot0 only works equal x / y resolutions, use mandelbrot() instead")
  } else {
    if (is.numeric(resolution) & length(resolution) == 1) {
      resolution <- as.integer(resolution)
      x_res <- resolution
      y_res <- resolution
    } else {
      stop("resolution should be an integer, not ", resolution)
    }
  }

  if (!is.numeric(xlim) | !is.numeric(ylim)) {
    stop("xlim and ylim must be numeric")
  }

  x_coord <- seq(xlim[1], xlim[2], len = x_res)
  y_coord <- seq(ylim[1], ylim[2], len = y_res)
  set <- numeric(x_res * y_res)

  # This is the call to the C function itself
  this_set <- .C("mandelbrot_alt",
    xcoo = as.double(x_coord),
    ycoo = as.double(y_coord),
    nx = as.integer(x_res),
    ny = as.integer(y_res),
    set = as.integer(set),
    iter = as.integer(iterations))$set

  df <- data.frame(expand.grid(x_coord, y_coord), this_set)
  colnames(df) <- c("x", "y", "value")
  df
}


src/mandelbrot.c
#include <stdio.h>

/***********************************************************
 *  Function to evaluate a series of complex numbers that  *
 *  form a subset of the Argand plane, to see if           *
 *  they belong to the Mandelbrot set.                     *
 *  This function has been written with the intention of   *
 *  linking it to some R code in order to create a         *
 *  practical way to play with the set.                    *
 *                                                         *
 *  Written by Mario dos Reis. September 2003              *
 *                                                         *
 ***********************************************************/

void mandelbrot_(double *xcoo, double *ycoo, int *nx,
  int *ny, int *set, int *iter)

  /* 'xcoo' and 'ycoo' are the x and y coordinates respectively
   *  of all the points to be evaluated.
   * 'nx' and 'ny' number of divisions along the x and y axes
   * 'set' will store the practical output of the function
   * 'iter' is the maximun number of iterations
   */

{
  int i, j, k;
  double z[2], c[2], oldz[2];

  for(i = 0; i < *nx; i++) {

    for(j = 0; j < *ny; j++) {

      /* initialise the complex point to be tested
       * c[0] (or z[0]) is the real part
       * c[1] (or z[1]) is the imaginary part
       */
      c[0] = xcoo[i]; c[1] = ycoo[j];
      z[0] = 0;       z[1] = 0;

      for(k = 1; k < *iter + 1; k++) {

        oldz[0] = z[0]; oldz[1] = z[1];

        // the mandelbrot mapping z -> z^2 + c
        z[0] = oldz[0]*oldz[0] - oldz[1]*oldz[1] + c[0];
        z[1] = 2 * oldz[0]*oldz[1] + c[1];

        if((z[0]*z[0] + z[1]*z[1]) > 4) {
          break;
        }
      }

      /* fills the set vector
       * notice the trick 'i * (*ny) + j' to find
       * the appropiate position of the output in the
       * vector set. The R function will take care of
       * transforming this vector into a suitable matrix
       * for plotting, etc.
       */
      set[i * (*ny) + j] = k;
    }
  }
  return;
}

/* Alt interface to above that generates set data in
 * a more convenient shape for fast reformatting
 */
void mandelbrot_alt(double *xcoo, double *ycoo, int *nx,
  int *ny, int *set, int *iter)

{
  int i, j, k;
  double z[2], c[2], oldz[2];

  for(i = 0; i < *nx; i++) {

    for(j = 0; j < *ny; j++) {

      c[0] = xcoo[i]; c[1] = ycoo[j];
      z[0] = 0;       z[1] = 0;

      for(k = 1; k < *iter + 1; k++) {

        oldz[0] = z[0]; oldz[1] = z[1];

        // the mandelbrot mapping z -> z^2 + c
        z[0] = oldz[0]*oldz[0] - oldz[1]*oldz[1] + c[0];
        z[1] = 2 * oldz[0]*oldz[1] + c[1];

        if((z[0]*z[0] + z[1]*z[1]) > 4) {
          break;
        }
      }

      set[j * (*nx) + i] = k;
    }
  }
  return;
}



R/util.R

#' Generate palette suitable for coloring a set
#'
#' Takes a simple palette and expands / oscillates
#' it for use with Mandelbrot sets.
#'
#' @param palette vector of color hex strings (e.g. '#FFFFFF')
#' @param fold wrap or fold the palette back on itself
#' @param reps number of times to replicate the color vector
#' @param in_set color for areas in the Mandelbrot set
#'
#' @return an extended color vector
#'
#' @examples
#' view <- mandelbrot(xlim = c(-0.8438146, -0.8226294),
#'   ylim = c(0.1963144, 0.2174996), iter = 500)
#'
#' # can be used to simply interpolate a color gradient
#' spectral <- RColorBrewer::brewer.pal(11, "Spectral")
#' cols <- mandelbrot_palette(spectral, fold = FALSE)
#' plot(view, col = cols, transform = "inv")
#'
#' # simple palettes might need folds / reps to look good
#' blues <- RColorBrewer::brewer.pal(9, "Blues")
#' cols <- mandelbrot_palette(blues, in_set = "white",
#'   fold = TRUE, reps = 2)
#' plot(view, col = cols, transform = "log")
#'
#' @export
mandelbrot_palette <- function(palette, fold = TRUE,
  reps = 1L, in_set = "black") {

  if (!fold %in% c(TRUE, FALSE)) {
    stop("Fold must be TRUE or FALSE")
  }

  if (length(palette) < 100) {
    palette <- grDevices::colorRampPalette(palette)(100)
  }

  if (fold) {
    palette <- c(palette, rev(palette))
  }

  c(rep(palette, reps), in_set)
}

#' Plot a Mandelbrot set using base graphics
#'
#' Draws colored set membership using \code{image}.
#'
#' @param x an object generated by \code{\link[mandelbrot]{mandelbrot}}
#' @param col a vector of colors, such as those generated by
#'   \code{\link[mandelbrot]{mandelbrot_palette}}
#' @param transform the name of a transformation to apply to the number
#'   of iterations matrix
#' @param asp the \code{asp} parameter to \code{\link[graphics]{image}}
#'   which controls aspect ratio
#' @param ... extra arguments passed to \code{\link[graphics]{image}}
#'
#' @importFrom grDevices grey.colors
#'
#' @export
plot.mandelbrot <- function(x,
  col = mandelbrot_palette(c("white", grey.colors(50))),
  transform = c("none", "inverse", "log"), asp = 1, ...) {

  transform <- match.arg(transform)
  old_par <- par()

  par(mar = rep(1, 4))

  if (transform != "none") {
    if (transform == "inverse") {
      x$z <- 1/x$z
      col <- rev(col)
    } else {
      if (transform == "log") {
        x$z <- log(x$z)
      } else {
        stop("transform not recognised: ", transform)
      }
    }
  }

  graphics::image(x, col = col, axes = FALSE, asp = asp, ...)

  par <- old_par
}

#' @export
print.mandelbrot <- function(x, ...) {
  cat(" Mandelbrot set view object within limits x:",
    paste(range(x$x), collapse = ", "),
    "and y:",
    paste(range(x$y), collapse = ", "),
    "\n")
  cat(" Iterations matrix:",
    paste(dim(x$z), collapse = " x "))

   invisible(x)
}

#' Convert Mandelbrot object to data.frame for plotting
#'
#' Converts objects produced by \code{\link[mandelbrot]{mandelbrot}}
#' to tidy data.frames for use with ggplot and other tidyverse packages.
#'
#' @param x a Mandelbrot set object produced by \code{\link[mandelbrot]{mandelbrot}}
#' @param ... ignored
#'
#' @return a 3-column \code{data.frame}
#'
#' @examples
#'
#' mb <- mandelbrot()
#' df <- as.data.frame(mb)
#' head(df)
#'
#' @export
as.data.frame.mandelbrot <- function(x, ...) {
  df <- reshape2::melt(x$z)
  #df <- data.table::melt(mandelbrot$z)
  df$x <- x$x[df$Var1]
  df$y <- x$y[df$Var2]
  df[,c("x", "y", "value")]
}

